src/OpenIddict.Core/OpenIddictBuilder.cs

```diff
@@ -1,19 +1,404 @@
+        /// Initializes a new instance of <see cref="OpenIddictBuilder"/>.
+        /// </summary>
+        /// <param name="services">The services collection.</param>
+        public OpenIddictBuilder(IServiceCollection services) {
+            Services = services;
+        }
+
+        /// <summary>
+        /// Gets or sets the type corresponding to the Application entity.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type ApplicationType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type corresponding to the Authorization entity.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type AuthorizationType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type corresponding to the Scope entity.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type ScopeType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type corresponding to the Token entity.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type TokenType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type corresponding to the User entity.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type UserType { get; set; }
+
+        /// <summary>
+        /// Gets the services used by OpenIddict.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public IServiceCollection Services { get; }
+
+        /// <summary>
+        /// Amends the default OpenIddict configuration.
+        /// </summary>
+        /// <param name="configuration">The delegate used to configure the OpenIddict options.</param>
+        /// <remarks>This extension can be safely called multiple times.</remarks>
+        /// <returns>The <see cref="OpenIddictBuilder"/>.</returns>
+        public virtual OpenIddictBuilder Configure([NotNull] Action<OpenIddictOptions> configuration) {
+            if (configuration == null) {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            Services.Configure(configuration);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Replaces the default application manager by a custom manager.
+        /// </summary>
+        /// <typeparam name="TManager">The type of the custom manager.</typeparam>
+        /// <returns>The <see cref="OpenIddictBuilder"/>.</returns>
+        public virtual OpenIddictBuilder ReplaceApplicationManager<TManager>() {
+            var contract = typeof(OpenIddictApplicationManager<>).MakeGenericType(ApplicationType);
+            if (!contract.IsAssignableFrom(typeof(TManager))) {
+                throw new InvalidOperationException("Custom managers must be derived from OpenIddictApplicationManager.");
+            }
```

Comment:
This piece of code to ensure the user provided type conforms to the expectation is duplicated, can be refactored into a private method with a couple of type parameters